atc/worker/fetcher.go

```diff
@@ -0,0 +1,170 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"io"
+	"time"
+
+	"github.com/concourse/concourse/atc/runtime"
+
+	"github.com/concourse/concourse/atc/resource"
+
+	"code.cloudfoundry.org/clock"
+	"code.cloudfoundry.org/lager"
+	"github.com/concourse/concourse/atc/db"
+	"github.com/concourse/concourse/atc/db/lock"
+)
+
+const GetResourceLockInterval = 5 * time.Second
+
+var ErrFailedToGetLock = errors.New("failed to get lock")
+var ErrInterrupted = errors.New("interrupted")
+
+//go:generate counterfeiter . Fetcher
+
+type Fetcher interface {
+	Fetch(
+		ctx context.Context,
+		logger lager.Logger,
+		containerMetadata db.ContainerMetadata,
+		gardenWorker Worker,
+		containerSpec ContainerSpec,
+		processSpec runtime.ProcessSpec,
+		resource resource.Resource,
+		owner db.ContainerOwner,
+		imageFetcherSpec ImageFetcherSpec,
+		cache db.UsedResourceCache,
+		lockName string,
+	) (GetResult, Volume, error)
+}
+
+func NewFetcher(
+	clock clock.Clock,
+	lockFactory lock.LockFactory,
+	fetchSourceFactory FetchSourceFactory,
+) Fetcher {
+	return &fetcher{
+		clock:              clock,
+		lockFactory:        lockFactory,
+		fetchSourceFactory: fetchSourceFactory,
+	}
+}
+
+type fetcher struct {
+	clock              clock.Clock
+	lockFactory        lock.LockFactory
+	fetchSourceFactory FetchSourceFactory
+}
+
+func (f *fetcher) Fetch(
+	ctx context.Context,
+	logger lager.Logger,
+	containerMetadata db.ContainerMetadata,
+	gardenWorker Worker,
+	containerSpec ContainerSpec,
+	processSpec runtime.ProcessSpec,
+	resource resource.Resource,
+	owner db.ContainerOwner,
+	imageFetcherSpec ImageFetcherSpec,
+	cache db.UsedResourceCache,
+	lockName string,
+) (GetResult, Volume, error) {
+	result := GetResult{}
+	var volume Volume
+	// TODO: resource_instance_fetch_source.go already knows which volume to use for the resource output, can this be consolidated
+	containerSpec.Outputs = map[string]string{
+		"resource": processSpec.Args[0],
+	}
+
+	// TODO: just pass in imageFetcherSpec not its contents
+	//		 this might be a bad idea, don't want the fetchsource to know about images?
+	fetchSource := f.fetchSourceFactory.NewFetchSource(
+		logger,
+		gardenWorker,
+		owner,
+		cache,
+		resource,
+		imageFetcherSpec.ResourceTypes,
+		containerSpec,
+		processSpec,
+		containerMetadata,
+		imageFetcherSpec.Delegate,
+	)
+
+	ticker := f.clock.NewTicker(GetResourceLockInterval)
+	defer ticker.Stop()
+
```

Comment:
Can we move ticker and it's defer after the error check return (just before the for loop) 
```
106 if err == nil || err != ErrFailedToGetLock {
107		return result, volume, err
108	}
```

Reply:
done